sdk/python/feast/registry.py

```diff
@@ -0,0 +1,226 @@
+                registry_proto.feature_tables
+            ):
+                if (
+                    existing_feature_table_proto.spec.name == name
+                    and existing_feature_table_proto.project == project
+                ):
+                    del registry_proto.feature_tables[idx]
+                    return registry_proto
+            raise Exception(f"Feature table {name} does not exist in project {project}")
+
+        self._registry_store.update_registry(updater)
+        return
+
+
+class RegistryStore(ABC):
+    @abstractmethod
+    def get_registry(self):
+        pass
+
+    @abstractmethod
+    def update_registry(self, updater: Callable[[RegistryProto], RegistryProto]):
+        pass
+
+
+class LocalRegistryStore(RegistryStore):
+    def __init__(self, filepath: str):
+        self._filepath = Path(filepath)
+        if not self._filepath.exists():
+            registry_proto = RegistryProto()
+            registry_proto.registry_schema_version = REGISTRY_SCHEMA_VERSION
+            self._write_registry(registry_proto)
+        return
+
+    def get_registry(self):
+        registry_proto = RegistryProto()
+        registry_proto.ParseFromString(self._filepath.read_bytes())
+        return registry_proto
+
+    def update_registry(self, updater: Callable[[RegistryProto], RegistryProto]):
+        registry_proto = self.get_registry()
+        registry_proto = updater(registry_proto)
+        self._write_registry(registry_proto)
+        return
+
+    def _write_registry(self, registry_proto: RegistryProto):
+        registry_proto.version_id = str(uuid.uuid4())
+        registry_proto.last_updated.FromDatetime(datetime.utcnow())
+        self._filepath.write_bytes(registry_proto.SerializeToString())
+        return
+
+
+class GCPRegistryStore(RegistryStore):
+    def __init__(self, uri: str):
+        try:
+            from google.cloud import storage
+            from google.cloud.exceptions import NotFound
+        except ImportError:
+            raise ImportError(
+                "Install package google-cloud-storage==1.20.* for gcs support"
+                "run ```pip install google-cloud-storage==1.20.*```"
+            )
+        try:
+            self.gcs_client = storage.Client(project="kf-feast")
+        except DefaultCredentialsError:
+            self.gcs_client = storage.Client.create_anonymous_client()
+        self._uri = urlparse(uri)
+        self._bucket = self._uri.hostname
+        self._blob = self._uri.path.lstrip("/")
+        try:
+            bucket = self.gcs_client.get_bucket(self._bucket)
+        except NotFound:
+            bucket = self.gcs_client.create_bucket(self._bucket)
+            print(bucket)
+        if not storage.Blob(bucket=bucket, name=self._blob).exists(self.gcs_client):
+            registry_proto = RegistryProto()
+            registry_proto.registry_schema_version = REGISTRY_SCHEMA_VERSION
+            self._write_registry(registry_proto)
```

Comment:
I think it's a bit of a trade off on when we would try and read/write to the registry. My intuition is to do this only when the user runs the `apply()` commands. Does writing the proto here mean that when the user creates an instance of `Client`, we will start writing to the object store?